src/charm.py

```diff
@@ -82,91 +72,47 @@ def __init__(self, *args):
             source_url=self._external_http_server_url,
             refresh_event=[
                 # refresh the source url when TLS config might be changing
-                self.on[self.cert_handler.certificates_relation_name].relation_changed,
+                self.on[self.coordinator.cert_handler.certificates_relation_name].relation_changed,
                 # or when ingress changes
                 self.ingress.on.ready,
             ],
         )
-        # # Patch the juju-created Kubernetes service to contain the right ports
-        self.unit.set_ports(*self.tempo.all_ports.values())
-
-        # Provide ability for Tempo to be scraped by Prometheus using prometheus_scrape
-        self._scraping = MetricsEndpointProvider(
-            self,
-            relation_name="metrics-endpoint",
-            jobs=[{"static_configs": [{"targets": [f"*:{tempo.tempo_http_server_port}"]}]}],
-        )
-        self._grafana_dashboards = GrafanaDashboardProvider(
-            self, relation_name="grafana-dashboard"
-        )
 
         self.tracing = TracingEndpointProvider(self, external_url=self._external_url)
-        self._inconsistencies = self.coordinator.get_deployment_inconsistencies(
-            has_s3=self.s3_ready
-        )
-        self._is_consistent = not self._inconsistencies
 
         # We always listen to collect-status
         self.framework.observe(self.on.collect_unit_status, self._on_collect_unit_status)
 
-        if not self._is_consistent:
-            logger.error(
-                f"Inconsistent deployment. {self.unit.name} will be shutting down. "
-                "This likely means you need to add an s3 integration. "
-                "This charm will be unresponsive and refuse to handle any event until "
-                "the situation is resolved by the cloud admin, to avoid data loss."
-            )
-            return  # refuse to handle any other event as we can't possibly know what to do.
+        # refuse to handle any other event as we can't possibly know what to do.
+        if (
+            not self.coordinator.cluster.has_workers
+            or not self.coordinator.is_coherent
+            or not self.coordinator.s3_ready
+        ):
+            return
```

Comment:
I assume we're not logging anything here because the coordinator will do that for us under the same circumstances? (let's note this in a comment)

if so, we should consider exposing this conditions from the coordinator. 

How about a `Coordinator.can_handle_events --> self.has_workers & self.is_coherent & self.s3_ready`?

@lucabello

Reply:
I think it's a good idea! I have a PR open in cos-lib, I'll sneak this in!